### PR TITLE
ICU-22520 Refactor function macros into inline functions

### PR DESCRIPTION
**Pull request by roubert**
This is to facilitate further refactoring of the locale code.

##### Checklist

- [X] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22520
- [X] Required: The PR title must be prefixed with a JIRA Issue number.
- [X] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [X] Required: Each commit message must be prefixed with a JIRA Issue number.
- [X] Issue accepted (done by Technical Committee after discussion)
- [ ] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

File: icu4c/source/common/loclikely.cpp [hunk 1]
```diff
@@ -199,24 +199,27 @@ parseTagString(
...
+inline bool CHECK_TRAILING_VARIANT_SIZE(const char* trailing, int32_t trailingLength) {
+    int32_t count = 0;
+    int32_t i;
+    for (i = 0; i < trailingLength; i++) {
+        if (trailing[i] == '-' || trailing[i] == '_') {
+            count = 0;
+            if (count > 8) {
+                return false;
+            }
+        } else if (trailing[i] == '@') {
```
Comment (markusicu) [inline icu4c/source/common/loclikely.cpp:212]: just FYI: I see that we already have fairly low-level code that won't work on EBCDIC machines...

File: icu4c/source/common/loclikely.cpp [hunk 2]
```diff
@@ -266,7 +269,8 @@ _uloc_addLikelySubtags(const char* localeID,
     trailing = &localeID[trailingIndex];
     trailingLength = (int32_t)uprv_strlen(trailing);
 
-    CHECK_TRAILING_VARIANT_SIZE(trailing, trailingLength);
+    if (!CHECK_TRAILING_VARIANT_SIZE(trailing, trailingLength)) goto error;
```
Comment (markusicu) [inline icu4c/source/common/loclikely.cpp:272]:
```suggestion
    if (!CHECK_TRAILING_VARIANT_SIZE(trailing, trailingLength)) { goto error; }
```
... and can we do better now than `goto error`?

File: icu4c/source/common/loclikely.cpp [hunk 3]
```diff
@@ -362,7 +366,7 @@ _uloc_minimizeSubtags(const char* localeID,
...
```
Comment (markusicu) [inline icu4c/source/common/loclikely.cpp:369]:
```suggestion
    if (!CHECK_TRAILING_VARIANT_SIZE(trailing, trailingLength)) { goto error; }
```

File: icu4c/source/common/loclikely.cpp (hunk 2 above)
Comment (roubert) [reply icu4c/source/common/loclikely.cpp:272]: I'll be coming for `goto` next, that's just easier to do if getting this out of macros first.

Comment (markusicu) [reply icu4c/source/common/loclikely.cpp:272]: ok, but braces please?

Comment (roubert) [reply icu4c/source/common/loclikely.cpp:272]: Is spending an hour on running the CI after applying that change really a meaningful use of time for something that's about to be deleted anyway?

Comment (roubert) [reply icu4c/source/common/loclikely.cpp:272]: Applying those changes will trigger a CI run, which takes about an hour. It then leaves the PR with commits that'll need to be squashed before the PR can be merged. Squashing them will trigger a CI run, which takes about an hour again. Is that really a meaningful use of time for something that's about to be replaced anyway?

Comment (markusicu) [reply icu4c/source/common/loclikely.cpp:272]: amending the commit would trigger only one round, not two...

**Review by markusicu — APPROVED** [review]

**Pull request merged**